packages/astro/test/fixtures/middleware-dev/src/pages/404.astro

```diff
@@ -0,0 +1,6 @@
+---
```

Comment:
If 404.astro exists, this issue might be fixing #8154, **not** #7881. I think the older issue may even be intended behavior.

Reply:
@lilnasy why do you think #7881 should be the intended behaviour? Genuinely asking! You might be onto something

Reply:
I'm thinking about the cases where an adapter tries to handle unmatched requests by bypassing the core app.

Some serverless platforms may not even call `app.render()`: if they find no matching route and no prerendered page, they might fallback to a platform-specific 404 page.

On a different note, if we were to wrap middleware around our default response (`new Response(null, { status: 404 })`), it would likely cause another wave of issues because I doubt many users are accounting for `next()`'s body being null.

Reply:
That makes total sense! I agree with every point.